File: ios/CryptoBridge.m
#import "CryptoBridge.h"
#import "BCDerivation.h"
#import "BCCrypto.h"

@implementation CryptoBridge

RCT_EXPORT_MODULE(CryptoBridge)

RCT_REMAP_METHOD(deriveKeyFromPassword,
                 deriveKeyFromPassword:(NSString *)password
                 andSalt:(NSString *)salt
                 forRounds:(int)rounds
                 withBits:(int)bits
                 withResolver:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)
{
    NSString *result = [BCDerivation deriveKeyFromPassword:password andSalt:salt forRounds:rounds withBits:bits];
    resolve(result);
}

RCT_REMAP_METHOD(generateSaltWithLength,
                 generateSaltWithLength:(int)length
                 withResolver:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)
{
    NSString *result = [BCCrypto generateSaltWithLength:length];
    resolve(result);
}

RCT_REMAP_METHOD(generateIV,
                 withResolver:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)
{
    NSString *result = [BCCrypto generateIVHex];
    resolve(result);
}

RCT_REMAP_METHOD(encryptText,
                 encryptText:(NSString *)text
                 withKey:(NSString *)key
                 andSalt:(NSString *)salt
                 andIV:(NSString *)ivHex
                 andHMAC:(NSString *)hmacHexKey
                 withResolver:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)
{
    NSString *result = [BCCrypto encryptText:text withKey:key andSalt:salt andIV:ivHex andHMAC:hmacHexKey];
    resolve(result);
}

RCT_REMAP_METHOD(decryptText,
                 decryptText:(NSString *)encryptedText
                 withKey:(NSString *)key
                 andIV:(NSString *)ivHex
                 andSalt:(NSString *)saltHex
                 andHMACKey:(NSString *)hmacHexKey
                 andHMAC:(NSString *)hmacHex
                 withResolver:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)
{
    NSString *result = [BCCrypto decryptText:encryptedText withKey:key andIV:ivHex andSalt:saltHex andHMACKey:hmacHexKey andHMAC:hmacHex];
    if ([result rangeOfString:@"Error="].location == 0) {
        NSString *errorString = [result substringFromIndex:6];
        NSError *error = [NSError errorWithDomain:@"pw.buttercup.mobile" code:1 userInfo:@{
          @"error": errorString
        }];
        reject(@"error", [[NSString alloc] initWithFormat:@"Decryption failed: %@", errorString], error);
        return;
    }
    resolve(result);
}

@end

File: android/app/src/main/java/pw/buttercup/mobile/CryptoBridge.java
package pw.buttercup.mobile;

import com.facebook.react.bridge.NativeModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;

import java.net.URLDecoder;
import java.lang.String;
import java.util.Map;
import java.util.HashMap;

public class CryptoBridge extends ReactContextBaseJavaModule {

    private String join(String[] items) {
        String output = "";
        for (String item : items) {
            if (output.length() > 0) {
                output += ",";
            }
            output += item;
        }
        return output;
    }

    public CryptoBridge(ReactApplicationContext reactContext) {
        super(reactContext);
    }

    @ReactMethod
    public void decryptText(String encryptedText, String keyHex, String ivHex, String saltHex, String hmacHexKey, String hmacHex, Promise promise) {
        String decryptedText = BCCrypto.decryptText(encryptedText, keyHex, ivHex, saltHex, hmacHexKey, hmacHex);
        promise.resolve(decryptedText);
    }

    @ReactMethod
    public void deriveKeyFromPassword(String password, String salt, int rounds, int bits, Promise promise) {
        try {
            String derivedKey = BCDerivation.deriveKeyFromPassword(password, salt, rounds, bits);
            promise.resolve(derivedKey);
        } catch (Exception e) {
            promise.reject("Failed deriving key: " + e.getMessage(), e);
        }
    }

    @ReactMethod
    public void encryptText(String encodedText, String keyHex, String saltHex, String hmacHexKey, Promise promise) {
        String text;
        try {
            // We decode the text from Base64 and URI-encoding due to limitations with the bridge
            // sending weird text that resulted from GZIP'ing:
            text = URLDecoder.decode(BCHelpers.base64ToString(encodedText), "UTF-8");
        } catch (Exception ex) {
            promise.reject("Failed encrypting data: " + ex.getMessage(), ex);
            return;
        }
        String encryptedText = BCCrypto.encryptText(text, keyHex, saltHex, hmacHexKey);
        promise.resolve(encryptedText);
    }

    @ReactMethod
    public void generateIV(Promise promise) {
        String iv;
        try {
            iv = BCCrypto.generateIV();
        } catch (Exception ex) {
            promise.reject("Failed generating IV: " + ex.getMessage(), ex);
            return;
        }
        promise.resolve(iv);
    }

    @ReactMethod
    public void generateSaltWithLength(int length, Promise promise) {
        String salt = BCCrypto.generateSalt(length);
        promise.resolve(salt);
    }

    @Override
    public String getName() {
        return "CryptoBridge";
    }

}
